### DISCUSSION TITLE
how to modify topics and make it work in consumer at runtime

### DISCUSSION THREAD

**Discussion by moonService** [spring-projects/spring-kafka#2371] [Q&A]
if topicPattern not enough to describe the possible changes in the topics, how to modify topics and make it work in consumer at runtime

**Comment by garyrussell** [comment]
It is not clear what you are asking. Perhaps you are referring to this recent enhancement:

https://github.com/spring-projects/spring-kafka/blob/3d8dcce580ce1752867a8833b0544be4fb53acd3/spring-kafka/src/main/java/org/springframework/kafka/core/KafkaAdmin.java#L146-L154

??

**Reply by artembilan** [reply]
I guess the request is more about subscribing to new topics at runtime.
I wonder if we can (or if that makes sense at all) to implement some `add(remove)Topics()` into a `ConcurrentMessageListenerContainer` and restart it respectively with a subsequent rebalance and so on...

**Reply by garyrussell** [reply]
You can use a pattern (regex) to specify the topics and they will be picked up dynamically (checks are every 5 minutes by default).

```java
/**
* Create properties for a container that will subscribe to topics matching the
* specified pattern. The framework will create a container that subscribes to all
* topics matching the specified pattern to get dynamically assigned partitions. The
* pattern matching will be performed periodically against topics existing at the time
* of check.
* @param topicPattern the pattern.
* @see org.apache.kafka.clients.CommonClientConfigs#METADATA_MAX_AGE_CONFIG
*/
public ContainerProperties(Pattern topicPattern) {
```

**Reply by moonService** [reply]
@artembilan that's i needed

**Reply by moonService** [reply]
@garyrussell we can't use pattern (regex) to describe all changes what is possible

**Reply by garyrussell** [reply]
We can investigate it; stopping/starting the container after altering the topics would be the simplest solution, but that would cause a rebalance. If that's acceptable, it's probably not too hard to do.

Changing them at runtime would be much more difficult because it would need to be robust with many test cases.
